### ISSUE TITLE
[TextView] [CSS] text-transform support?

### ISSUE THREAD

**Issue by grabbou** [facebook/react-native#2088]
Hey,
Given recent updates to `text-decoration` implementation, I was wondering if this would've been in scope of React Native? I know we can pretty easily do this:
```
<Text>{'something'.toUpperCase()}</Text>
```
but having universal apps in mind, basic support for `upperCase` and `lowerCase` would be nice.
```
<Text style={{textTransform: 'uppercase'}}>something</Text>
```

Depending on how Text is rendered, either basic NSString bindings below might be used or plain JS methods:
- `capitalize`  - NSString -capitalizedString
- `uppercase` - NSString -uppercaseString
- `lowercase` - NSString -lowercaseString

If the answer to my question is Yes - I am happy to add in this in a PR.

**Comment by sahrens** [comment]
Not sure I understand the request - what do you want exactly that you can't do with plain JavaScript?

**Comment by grabbou** [comment]
Well, just wondering what are the pros/cons of implementing `textTransform` so we don't have to use plain JS (it's mainly for the sake of CSS-like approach as I am happy with plain JS anyway). Not a real issue, feel free to close if no point in implementing.

**Comment by sahrens** [comment]
Our general philosophy on RN is not to provide APIs unless they both
provide clear value and can't be reasonably done otherwise. Since you can
easily use JS to do this, I don't think we'll provide specific support for
it.

**Comment by brentvatne** [comment]
Adding on to @sahrens' comment above, if anyone comes across some use cases where there is a significant benefit to doing this in a style rather than with JS, please post them here so we can consider!

**Comment by ide** [comment]
+1 strongly in favor of keeping code in user space.

**Comment by mikefrancis** [comment]
Just stumbled upon this thread from Google; I'd say that adding `textTransform` support is necessary to separate your concerns. As others have pointed out, you _can_ do this in JS, however what's the point in having a class full of styles and then having to use JS to force caps.

Say a couple of months down the line you don't wish to use caps any more then you've got to trawl through your app removing all the references to it. Having this as a style would alleviate that problem.

**Comment by jonlong** [comment]
@brentvatne Just dropping in to echo @mikefrancis's comments, which are right on the money in my experience.

Here's a quick example from a project I'm working on. This is the same screen with two different design options — labels rendered in title case, or labels rendered in small caps.

![texttransform](https://cloud.githubusercontent.com/assets/482527/13441988/20705db4-dfc8-11e5-8a16-d4926a207069.png)

There's two things that make a `textTransform` property attractive to me in this case. The first is that I think there's a semantic difference between setting the actual content of `<Text>` to `CITY` rather than `City`. If preserving that semantic state is important, then I think this becomes a matter of style.

The second is to @mikefrancis's points around separation of concerns and maintainability. If it's a matter of style, then it makes sense to manage it with stylesheets. I see it as sort of analogous to inline styles in HTML — they'll work, but if the rest of the app uses stylesheets, it's a lot less flexible to have presentational styles mixed into markup.

Here's a sample of the JSX for both of those views:

**Title Case**: 

```jsx
<View { ...style('field') }>
  <TextInput
    { ...style('input') }
  />
  <Text { ...style('label') }>Address Line 1</Text>
</View>

<View { ...style('field') }>
  <TextInput
    { ...style('input') }
  />
  <Text { ...style('label') }>Address Line 2</Text>
</View>

<View { ...style('field') }>
  <TextInput
    { ...style('input') }
  />
  <Text { ...style('label') }>City</Text>
</View>

<View { ...style('field') }>
  <TextInput
    { ...style('input') }
  />
  <Text { ...style('label') }>State</Text>
</View>

<View { ...style('field') }>
  <TextInput
    { ...style('input') }
  />
  <Text { ...style('label') }>Zipcode</Text>
</View>
```

And the same code modified for **small caps**:

```jsx
<View { ...style('field') }>
  <TextInput
    { ...style('input') }
  />
  <Text { ...style('label.upper') }>{'Address Line 1'.toUpperCase()}</Text>
</View>

<View { ...style('field') }>
  <TextInput
    { ...style('input') }
  />
  <Text { ...style('label.upper') }>{'Address Line 2'.toUpperCase()}</Text>
</View>

<View { ...style('field') }>
  <TextInput
    { ...style('input') }
  />
  <Text { ...style('label.upper') }>{'City'.toUpperCase()}</Text>
</View>

<View { ...style('field') }>
  <TextInput
    { ...style('input') }
  />
  <Text { ...style('label.upper') }>{'State'.toUpperCase()}</Text>
</View>

<View { ...style('field') }>
  <TextInput
    { ...style('input') }
  />
  <Text { ...style('label.upper') }>{'Zipcode'.toUpperCase()}</Text>
</View>
```

In the second version, a simple design change requires a lot of updates to the template content itself. Even going the "Find and Replace" route would be pretty tricky.

I also find that in many scenarios, small caps tend to look better with a little bit of `letterSpacing` applied. In that case, I'd also end up modifying my styles too.

IMO, it's a lot of overhead for a common use case, and something that could go from all this to a single line in a stylesheet with `textTransform` support.

**Comment by brentvatne** [comment]
Another option is to wrap the Text component with one that adds the capabilities that you want, so then if you have the uppercase prop set on it, it will `toUpperCase()` everything. Can you create an issue on ProductPains that includes the great description of the problem that you just stated @jonlong?

**Comment by jonlong** [comment]
@brentvatne sure thing — done: https://productpains.com/post/react-native/texttransform-should-be-available-in-stylesheet/

**Comment by af7** [comment]
+1 I really need this on a specific app that uses it a lot. Thanks!

**Comment by dwilt** [comment]
I'm in a situation where I can't access the `Text` tags directly that I want to put uppercase because I'm dynamically generating them - thus, applying conditional JS is much harder than just passing in styles.

**Comment by vanpoos** [comment]
+1 for me as well. I am a designer which likes to develop UI's when I get the chance. Unfortunately, my Javascript skills are not great, yet. Having a more CSS centric approach to building UI's would really help me. Granted I should probably just learn Javascript and be done with it. However, I am sure there are others like me which like to code UI's but have not reached the level where they can use Javascript as an alternative. My 2 cents.

**Comment by ethanshar** [comment]
+1 it's clearly a style concern. it was a concern on the web and now on mobile.
I wouldn't want to worry about conditionally upper-casing text of a button component every time I use it but rather specify its style exactly like color or font-size, especially when I have to support two different style for the same component, each one with different text-transform style (Android vs iOs)

**Comment by leeight** [comment]
+1

**Comment by guiherzog** [comment]
+1

**Comment by neiker** [comment]
+1

**Comment by stephenmeszaros** [comment]
+1

**Comment by mjracca** [comment]
+1

**Comment by amazingandyyy** [comment]
the advantage of css "class" is to not duplicate codes.

I strongly think we need "whole supports" css system for RN.

**Comment by herlon214** [comment]
+1

**Comment by rstphnsn** [comment]
+1

**Comment by hbierlee** [comment]
+1

**Comment by jonlong** [comment]
For all the folks posting "+1" feedback in this thread, consider that it triggers a notification (often an email) not just for the React Native team, but for anyone subscribed to the thread (which can easily be several hundred people at once).

On behalf of those who'd like to keep tabs on this thread's progress without receiving daily "+1" emails, please consider using GitHub's reaction feature — it's a nice new way of registering +1-style feedback without blanket notifications to an entire subscriber list:

![image](http://g.recordit.co/rZsk5sgBU9.gif)

**Comment by manfredxu99** [comment]
+1

**Comment by alexpriceco** [comment]
So is this happening, or is it just closed and left to die? 😕

**Comment by adaptabi** [comment]
Why closing this? It seems so imature decision to not support obvious CSS properties and to duplicate the code for cross-universal apps!
This is a very easy to be implemented native but very hard to duplicate and take care of it in every single app.

**Comment by kenma9123** [comment]
+2

**Comment by sahrens** [comment]
It seems like there is enough desire here that if someone wants to submit a PR for this, we'd probably take it.

**Comment by rikur** [comment]
+1 We're building components that work in both web and react-native and using JS to transform text in web feels so wrong.

**Comment by BigFav** [comment]
Can this be reopened if a PR will be accepted?

It is a bit weird to have someone write up a request on ProductPains, state that a PR will likely be accepted, all while keeping the issue closed; it sends mixed messages, and since people use the issues to help decide what they want to work on it is pretty disingenuous.

**Comment by RafaelOda** [comment]
I agree with @mikefrancis and @jonlong opinions.
Another issue that I face is when you are dealing with optional text nodes, for example.

Doing something like 

```javascript

render() {
  let { thisMightBeNullOrText } = this.props;
  if (!!thisMightBeNullOrText) {
    return (
      <Text>{thisMightBeNullOrText}</Text>
    );
  } else {
    return null;
  }
}
```

seems worse than 

```javascript

render() {
  let { thisMightBeNullOrText } = this.props;
  return (
    <Text style={styles.toUpper}>{thisMightBeNullOrText}</Text>
  );
}
```

```javascript 
const styles = StyleSheet.create({
  toUpper: {
    textTransform: 'uppercase',
  },
})
```

**Comment by brandensilva** [comment]
> Our general philosophy on RN is not to provide APIs unless they both
provide clear value and can't be reasonably done otherwise. Since you can
easily use JS to do this, I don't think we'll provide specific support for
it.

@sahrens Providing clear value means you'd come to the conclusion that textTransform is far superior than inlining a toUppercase() all over <Text /> elements which quickly leads to unmaintainable code. With a pseudo CSS property it can be applied to multiple elements and imported anywhere you need it.

I think there is still a lot of interest and desire in adding this into the API. Why not accept a PR to do it?

**Comment by bradleywilkerson** [comment]
+1

**Comment by Kernael** [comment]
+1

**Comment by diegorodriguesvieira** [comment]
+1

**Comment by peteychuk** [comment]
+1

**Comment by encorex** [comment]
+1

**Comment by MBach** [comment]
+1

**Comment by grabbou** [comment]
@sahrens @brentvatne given the amount of interest this thread gets, do we want to accept such PR? If yes, happy to work on it. 

Otherwise, I guess we can lock the thread.

**Comment by twvdb** [comment]
+1

**Comment by nobso** [comment]
+1

**Comment by compulsions** [comment]
+1

**Comment by sahrens** [comment]
Yeah I'll take a PR.

**Comment by lucideer** [comment]
> Our general philosophy on RN is not to provide APIs unless they both provide clear value and can't be reasonably done otherwise. Since you can easily use JS to do this, I don't think we'll provide specific support for it.

This issue has been closed for a while I see, but I just discovered it via Google and I'm surprised to see what I would consider the most obvious use-case for this feature not mentioned. CSS `text-transform` allows you to copy the original text to the clipboard independent of its appearance. I can't see any way this can be easily done with JS. Perhaps with some event capturing intercepting the clipboard copy and altering the content, but this hardly seems straightforward.

**Comment by JuiceCrawl** [comment]
Another use case for textTransform is if you are using react intl's <FormattedMessage>. You must wrap the message in a <Text> to style anything in <FormattedMessage> - you can't use toUpperCase() with <FormattedMessage> and if you want to uppercase your text, there's nothing to do this.

**Comment by philtrep** [comment]
> Our general philosophy on RN is not to provide APIs unless they both provide clear value and can't be reasonably done otherwise.

Separation of concerns is important. It is counter intuitive to do this in JS.

**Comment by lefoy** [comment]
+1

**Comment by townmulti** [comment]
+1

**Comment by finestpixels** [comment]
+1

**Comment by zhangkaiyulw** [comment]
+1

**Comment by locolab** [comment]
+1

**Comment by TakaGoto** [comment]
+1

**Comment by brettburwell** [comment]
+1

**Comment by milexjaro** [comment]
+1

**Comment by DariusAf** [comment]
+1

**Comment by alfrednerstu** [comment]
To get capitalized text you need to make your own JavaScript function as there are no one natively.
So in that case textTransform would have clear value.

**Comment by adriantombu** [comment]
2 years later, another brick in the wall 😂

**Comment by iamtekeste** [comment]
I guess I will be using `toUppercase`.

**Comment by steveamaza** [comment]
So it's late 2017 and this still hasn't been agreed to? 🙄

**Comment by Inlesco** [comment]
So... @sahrens - any PRs that have been posted to implement this much-needed feature?

Like others mentioned, separaton of concerns, maintainability and DRY are extremely important points to solve here. 

Also, keep in mind that this helps when porting React web to RN as well, which is a very important aspect, too.

I don't see a reason why nobody in RN team could quickly throw a PR for this and get it merged. It's like, what, 3-4 hours of work max for someone familiar with the framework's and Android's / iOS inside-outs?

**Comment by Dygerati** [comment]
+1

**Comment by blainemuri** [comment]
It should be default behavior that React-Native at least has 1:1 coverage to the basic CSS functionality that people use. Yes this can be done with JS, but depending on the situation it can be a lot simpler to do this with CSS.

**Comment by zhangkaiyulw** [comment]
This is a daily thread, always exist, never close

**Comment by holden-caulfield** [comment]
+1. 
Can't believe why this is still not there after years of being asked. First time I've checked this I was working on a white label platform that would spawn several apps with different themes.... It has all of the theming separated in `style` objects properly and yet having texts changing case depending on the theme still needs to be dealt with on the React component itself, where it does NOT belong

**Comment by NySiya** [comment]
It is just a text-transform, but it takes more than 2 years ... 😂

**Comment by brentvatne** [comment]
@NySiya - strange that nobody has bothered to submit a high quality pull request to solve this problem in 2 years! this is open source, after all 🤔

**Comment by antonwintergerst** [comment]
Here's an elegant way to deal with this in ES6:
Extend the String object:
`Object.assign(String.prototype, {
  toTitleCase() {
    return this.toLowerCase().split(' ').map((word) => {
      return word.replace(word[0], word[0].toUpperCase());
    }).join(' ');
  },
});`
Then access this new method with:
`
// transforms 'some string' to 'Some String'
String('some string').toTitleCase();
`

**Comment by lucideer** [comment]
> *Here's an elegant way to deal with this in ES6*

I think you might be overlooking the problem here. Where would you propose to run this method? On render or on copy?

There are two steps in this process:

1. Displaying the text to the user in the desired format - this is very easy to do - I don't think anyone in this thread has a problem with this step.
2. Retaining the original text and providing the user/clipboard with access to this text - this is the hard problem which is non-trivial to do in RN.

As far as I can see, your method only addresses step 1 (which was already solved/simple)?

**Comment by MrLoh** [comment]
Another use case for `text-transform` that I just ran into is if you have nested `Text` elements to combine styles only the inner most text element has access to the string, so the outer ones can't do something like `toUppercase()`.

**Comment by rikur** [comment]
It's been about a year since my last +1, so here I go again:
+1!

**Comment by uditalias** [comment]
+1

**Comment by tmpfs** [comment]
TL;DR I've implemented this in [prism](https://github.com/fika-community/prism)

From a userspace point of view it is more complicated than you might think. We need to mutate children and then only text nodes (implementation is [here](https://github.com/fika-community/prism/blob/master/src/experimentalPlugins.js)).

To be able to deal with `textTransform` in style rules we cannot pass it directly to `StyleSheet.create()` (it will error) so it is dealt with as an [invariant](https://github.com/fika-community/prism#invariants).

I've looked a little at the RN code and it appears to me it would be non-trivial (although welcome if implemented!) as `setStyleAttributePreprocessor` quite rightly has very little knowledge of other parts of the code (how does `textTransform` translate to mutating children? - I'm not sure). My guess is that it is not so simple to implement at the moment.

There is an important caveat regarding `none` in the [textTransform docs](https://github.com/fika-community/prism#texttransform).

Maybe the last +1? We have something in userspace, let's use it.

**Comment by TomSwift** [comment]
I created this PR which implements a `textTransform` Text style for iOS.   https://github.com/facebook/react-native/pull/18387

**Comment by 0xc0d3r** [comment]
+1

**Comment by nmerchant** [comment]
I'me somewhat surprised that serious developers on this project have suggested using JS to modify text strings when `text-transform` is a pretty standard and commonly used style.  How is this still missing 2 years later?

**Comment by kenma9123** [comment]
+3

**Comment by SofiaChuquin** [comment]
+1

**Comment by jayporta** [comment]
+1

**Comment by philtrep** [comment]
@nmerchant Especially since there's a PR for it by @TomSwift . Maybe @sahrens can reevaluate?

**Comment by TomSwift** [comment]
@nmerchant @philtrep I believe we're really close to having my PR merged.  It's a slower process than I'd realized.

**Comment by nmerchant** [comment]
@TomSwift that's only half the solution though...still need Android support :( but nevertheless, bless you for taking this up.

**Comment by neiker** [comment]
![image](https://user-images.githubusercontent.com/688444/38729082-0efbf920-3f12-11e8-82d1-3a1bfc13de3a.png)

**Comment by leozzitowned** [comment]
+1

**Comment by xabree** [comment]
+1

**Comment by melihcaliskan** [comment]
+1

**Comment by gustavopradoreis** [comment]
+1

**Comment by havinhthai** [comment]
+1

**Comment by Access2emma** [comment]
+1

**Comment by Kazanz** [comment]
+1

**Comment by MrLoh** [comment]
I'm not sure this weird +1 posting will achieve anything, except alerting everyone who is still subscribed to this. There are emoji reactions on github now and since this has been closed it would probably make much more sense to open a new issue about android support for this. What's really needed is someone with the needed Android experience to create a pull request though.

**Comment by mrruby** [comment]
+1

**Comment by julialuk** [comment]
+1

**Comment by MrLoh** [comment]
@mrruby @julialuk if you want this to happen, help to bring forward a pull request, or sit in silence and subscribe to this issue. `+1` spam is not helping this to come forward at all, it will only lead to people unsubscribing from this issue.

**Comment by guruganeshm-prft** [comment]
+1
